zone/expedition.cpp

```diff
@@ -0,0 +1,2275 @@
+{
+}
+
+Expedition* Expedition::TryCreate(
+	Client* requester, DynamicZone& dynamiczone, ExpeditionRequest& request)
+{
+	if (!requester || !zone)
+	{
+		return nullptr;
+	}
+
+	// request parses leader, members list, and lockouts while validating
+	if (!request.Validate(requester))
+	{
+		LogExpeditionsModerate("[{}] request by [{}] denied", request.GetExpeditionName(), requester->GetName());
+		return nullptr;
+	}
+
+	if (dynamiczone.GetInstanceID() == 0)
+	{
+		dynamiczone.CreateInstance();
+	}
+
+	if (dynamiczone.GetInstanceID() == 0)
+	{
+		// live uses this message when trying to enter an instance that isn't ready
+		// we can use it as the client error message if instance creation fails
+		requester->MessageString(Chat::Red, DZ_PREVENT_ENTERING);
+		LogExpeditions("Failed to create a dynamic zone instance for expedition");
+		return nullptr;
+	}
+
+	std::string expedition_uuid = EQ::Util::UUID::Generate().ToString();
+
+	// unique expedition ids are created from database via auto-increment column
+	auto expedition_id = ExpeditionDatabase::InsertExpedition(
+		expedition_uuid,
+		dynamiczone.GetInstanceID(),
+		request.GetExpeditionName(),
+		request.GetLeaderID(),
+		request.GetMinPlayers(),
+		request.GetMaxPlayers()
+	);
+
+	if (expedition_id)
+	{
+		dynamiczone.SaveToDatabase();
+
+		auto expedition = std::unique_ptr<Expedition>(new Expedition(
+			expedition_id,
+			expedition_uuid,
+			dynamiczone,
+			request.GetExpeditionName(),
+			ExpeditionMember{ request.GetLeaderID(), request.GetLeaderName() },
+			request.GetMinPlayers(),
+			request.GetMaxPlayers()
+		));
+
+		LogExpeditions(
+			"Created [{}] [{}] instance id: [{}] leader: [{}] minplayers: [{}] maxplayers: [{}]",
+			expedition->GetID(),
+			expedition->GetName(),
+			expedition->GetInstanceID(),
+			expedition->GetLeaderName(),
+			expedition->GetMinPlayers(),
+			expedition->GetMaxPlayers()
+		);
+
+		expedition->SaveMembers(request);
+		expedition->SaveLockouts(request);
+
+		auto inserted = zone->expedition_cache.emplace(expedition_id, std::move(expedition));
+
+		inserted.first->second->SendUpdatesToZoneMembers();
+		inserted.first->second->SendWorldExpeditionUpdate(ServerOP_ExpeditionCreate); // cache in other zones
+		inserted.first->second->SendLeaderMessage(request.GetLeaderClient(),
+			Chat::System, EXPEDITION_AVAILABLE, { request.GetExpeditionName() });
+
+		if (!request.GetNotAllAddedMessage().empty())
+		{
+			Client::SendCrossZoneMessage(request.GetLeaderClient(), request.GetLeaderName(),
+				Chat::System, request.GetNotAllAddedMessage());
+		}
+
+		return inserted.first->second.get();
+	}
+
+	return nullptr;
+}
+
+void Expedition::CacheExpeditions(MySQLRequestResult& results)
+{
+	if (!results.Success() || !zone)
+	{
+		return;
+	}
+
+	std::vector<uint32_t> expedition_ids;
+	std::vector<uint32_t> instance_ids;
+	std::vector<std::pair<uint32_t, uint32_t>> expedition_character_ids;
+
+	using col = LoadExpeditionColumns::eLoadExpeditionColumns;
```

Comment:
Repositories might make it easier to interact with things like this; not going to ask you to swap because it could introduce bugs with data type translation in your code - but at least be aware of them

Repositories generate a struct that represents the underlying data model which hides away having to work with magic numbers etc.

https://eqemu.gitbook.io/server/in-development/developer-area/repositories